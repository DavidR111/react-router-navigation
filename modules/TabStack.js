/* @flow */
/* eslint no-duplicate-imports: 0 */

import { Component } from 'react'
import { withRouter } from 'react-router'
import { StateUtils } from 'react-navigation'
import type { RouterHistory, Location } from 'react-router'
import type { NavigationState, TabRendererProps, Tabs, TabRoute, TabProps } from './TypeDefinitions'
import StackUtils from './StackUtils'

type Props = RouterHistory & {
  children: Array<React$Element<TabProps>>,
  render: (props: TabRendererProps) => React$Element<any>,
  forceSync?: boolean,
}

type State = {
  navigationState: NavigationState<TabRoute>,
  tabs: Tabs,
  rootIndex: number,
  history: { [key: number]: Array<Location> },
}

class TabStack extends Component<void, Props, State> {

  props: Props
  state: State

  // Initialyze navigation state with initial history
  constructor(props: Props): void {
    super(props)
    // Build the tab stack ($FlowFixMe)
    const { children, entries, location } = props
    const tabs = StackUtils.build(children)
    // Get initial route
    const currentRoute = StackUtils.getRoute(tabs, location)
    if (!currentRoute) throw new Error('No route found !')
    // Build navigation state
    const routes = tabs.map((tab) => ({
      // $FlowFixMe
      key: StackUtils.createKey({ key: tab.key }),
      routeName: tab.path,
    }))
    const index = routes.findIndex(({ routeName }) => currentRoute.key === routeName)
    const navigationState = { index, routes }
    const rootIndex = props.index || 0
    // Initialyze cached history
    const history = {
      [index]: entries.slice(location.index),
    }
    // Save everything
    this.state = { navigationState, tabs, rootIndex, history }
  }

  // Listen all history events
  componentWillReceiveProps(nextProps): void {
    // Get current route ($FlowFixMe)
    const { location, entries } = nextProps
    const { navigationState: { routes, index }, tabs, rootIndex } = this.state
    // Get current tab
    const currentRoute = routes[index]
    const currentTab = StackUtils.get(tabs, currentRoute)
    // Get next tab
    const nextRoute = StackUtils.getRoute(tabs, location)
    if (!nextRoute) return
    const nextTab = StackUtils.get(tabs, nextRoute)
    const nextIndex = routes.findIndex(({ routeName }) => routeName === nextRoute.key)
    // Update navigation state
    if (
      currentTab && nextTab &&
      StackUtils.shouldUpdate(currentTab, nextTab, this.props, nextProps)
    ) {
      this.setState((state) => ({
        navigationState: StateUtils.jumpToIndex(
          state.navigationState,
          nextIndex,
        ),
      }))
    }
    // Update history
    if (nextRoute) {
      this.state.history[nextIndex] = entries.slice(
        rootIndex,
        nextProps.index + 1,
      )
    }
  }

  // Callback for when the current tab changes
  onRequestChangeTab = (index: number): void => {
    const entries = this.state.history[index]
    const tab = this.state.tabs[index]
    if (tab) {
      if (this.props.forceSync) {
        const n = this.state.rootIndex - (this.props.index || 0)
        this.props.go(n)
        this.props.replace(tab.path)
        if (entries) {
          entries
            .slice(this.state.rootIndex + 1)
            .forEach(({ pathname }) => {
              this.props.push(pathname)
            })
          this.props.replace(
            entries[Math.max(0, parseInt(entries.length - 1))].pathname
          )
        }
      } else {
        this.props.replace(tab.path)
      }
    }
  }

  // Render view
  render(): React$Element<any> {
    return this.props.render({
      ...this.state,
      onRequestChangeTab: this.onRequestChangeTab,
    })
  }

}

export default withRouter(TabStack)
