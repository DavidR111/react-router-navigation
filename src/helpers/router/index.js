/* @flow */

export { default as withRouter } from './withRouter'
export { default as routerReducer } from './routerReducer'
export { default as Router} from './Router'
