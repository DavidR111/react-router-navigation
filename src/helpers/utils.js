/* @flow */

import _ from 'lodash'
import type { NavigationState, NavigationRoute, NavigationScene } from './../types'

export function extractScenes(
  children: Array<React$Element<NavigationScene>>
): Array<NavigationRoute> {
  return children.map((child) => ({
    ...child.props,
    key: child.key,
  }))
}

export function normalizePath(path: string): string {
  return path
    .split('.')
    .filter((_path) => _path)
    .map((_path) => `routes[${_path}]`)
    .join('.')
}

export function getSiblingScenes(state: NavigationState) {
  const normalizedPath = normalizePath(state.path.slice(0, -2))
  const path = normalizedPath
    ? `${normalizedPath}.children`
    : 'children'
  const sceneChildren = _.get(state, path)
  return extractScenes(sceneChildren)
}

export function getCurrentRoute(state: NavigationState): NavigationRoute {
  const normalizedPath = normalizePath(state.path)
  return _.get(state, normalizedPath)
}

export function findPathOfClosestTabs(state: NavigationState): string {
  const paths = state.path.split('.')
  const path = paths
    .map((_path, index) => paths.slice(0, index + 1).join('.'))
    .reverse()
    .map((_path) => ({
      path: _path,
      navigationState: _.get(state, normalizePath(_path)),
    }))
    .filter(({ navigationState }) => navigationState.tabs)
    .map((tabState) => tabState.path)
    .join('')
  return path
}
