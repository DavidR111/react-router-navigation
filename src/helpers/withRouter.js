/* @flow */

import React, { PropTypes } from 'react'
import hoistStatics from 'hoist-non-react-statics'

function getDisplayName(WrappedComponent: () => React$Element<any>) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}

export default function withRouter(WrappedComponent: () => React$Element<any>) {
  const WithRouter = (props, context) => (
    <WrappedComponent
      {...props}
      {...context}
    />
  )
  WithRouter.contextTypes = {
    router: PropTypes.object.isRequired,
    navigationState: PropTypes.object.isRequired,
  }
  WithRouter.displayName = `withRouter(${getDisplayName(WrappedComponent)})`
  WithRouter.WrappedComponent = WrappedComponent
  return hoistStatics(WithRouter, WrappedComponent)
}
