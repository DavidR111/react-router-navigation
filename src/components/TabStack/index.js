/* @flow */

import React, { PropTypes, Component } from 'react'
import _ from 'lodash'
import CardStack from './../CardStack'
import TabView from './../TabView'
import { normalizePath } from './../../utils'
import type { NavigationState, NavigationSceneProps } from './../../types'

type Props = {
  navigationState: NavigationState,
  renderScene: (sceneProps: NavigationSceneProps) => React$Element<any>,
  renderHeader: (sceneProps: NavigationSceneProps) => React$Element<any> | null,
  push: (key: string, callback: Function) => void,
  pop: (callback: Function) => void,
  changeTab: (index: number, callback: Function) => void,
}

type DefaultProps = {
  renderHeader: () => null,
}

type State = {
  navigationState: NavigationState,
}

class TabStack extends Component {

  state: State
  props: Props

  static defaultProps: DefaultProps = {
    renderHeader: () => null,
  }


  // Clone navigation state to it's own
  // in this component and remove circular
  // references
  constructor(props: Props) {
    super(props)
    const { navigationState } = props
    const path = normalizePath(navigationState.path.slice(0, -4))
    const extractedNavigationState = path
      ? _.cloneDeep(_.get(navigationState, path))
      : navigationState
    this.state = {
      navigationState: extractedNavigationState,
    }
  }


  // Dispatch actions
  push = (key: string) => {
    this.props.push(key, this.updateNavigationState)
  }
  changeTab = (index: number) => {
    this.props.changeTab(index, this.updateNavigationState)
  }
  pop = () => {
    this.props.pop(this.updateNavigationState)
  }


  // Update local navigation state each
  // time an action is dispatched
  updateNavigationState = (navigationState: NavigationState): void => {
    const path = normalizePath(navigationState.path.slice(0, -4))
    const nextNavigationState = path
      ? _.cloneDeep(_.get(navigationState, path))
      : navigationState
    this.setState({ navigationState: nextNavigationState })
  }


  // Provided updated router context
  static childContextTypes = {
    router: PropTypes.object.isRequired,
  }
  getChildContext() {
    return {
      router: {
        push: this.push,
        pop: this.pop,
      },
    }
  }


  // Render each tab in a <CardStack />
  // component
  render() {
    return (
      <TabView
        navigationState={this.state.navigationState}
        changeTab={this.changeTab}
        renderScene={(navigationState) => (
          <CardStack
            navigationState={navigationState}
            pop={this.pop}
            renderScene={this.props.renderScene}
            renderHeader={(sceneProps) => this.props.renderHeader(sceneProps, this.pop)}
          />
        )}
      />
    )
  }

}

export default TabStack
