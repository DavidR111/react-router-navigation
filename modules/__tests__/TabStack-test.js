import React from 'react'
import { View } from 'react-native'
import { Match } from 'react-router'
import { TestRouter, componentFactory, TabView } from './helpers'
import TabStack from './../TabStack'
import renderer from 'react-test-renderer'

it('<TabStack /> renders correctly', () => {
  const component = renderer.create(
    <TestRouter>
      <TabStack render={TabView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </TabStack>
    </TestRouter>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> renders correctly by ignoring non <Match /> component', () => {
  const component = renderer.create(
    <TestRouter>
      <TabStack render={TabView}>
        <View pattern="/" />
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
      </TabStack>
    </TestRouter>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> renders correctly with initialEntries prop ', () => {
  const component = renderer.create(
    <TestRouter initialEntries={['/hello']}>
      <TabStack render={TabView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </TabStack>
    </TestRouter>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> re-renders correctly when "replace" action is called', () => {
  const component = renderer.create(
    <TestRouter>
      <TabStack render={TabView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </TabStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.props.onPress((history) => history.replace('/hello'))
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> re-renders correctly when onRequestChangeTab() method is called', () => {
  const component = renderer.create(
    <TestRouter>
      <TabStack render={TabView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </TabStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.children[0].props.onRequestChangeTab(2)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
