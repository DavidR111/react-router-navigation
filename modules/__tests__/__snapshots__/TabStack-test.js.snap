exports[`test <TabStack /> re-renders correctly when "replace" action is called 1`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 0,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <TabStack /> re-renders correctly when "replace" action is called 2`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 1,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;

exports[`test <TabStack /> re-renders correctly when onRequestChangeTab() method is called 1`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 0,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <TabStack /> re-renders correctly when onRequestChangeTab() method is called 2`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 2,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Goodbye
    </Text>
  </View>
</View>
`;

exports[`test <TabStack /> renders correctly 1`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 0,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <TabStack /> renders correctly with initialEntries prop  1`] = `
<View
  onPress={[Function]}>
  <View
    configureTransition={[Function]}
    initialLayout={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
    navigationState={
      Object {
        "index": 1,
        "routes": Array [
          Object {
            "key": "/",
          },
          Object {
            "key": "/hello",
          },
          Object {
            "key": "/goodbye",
          },
        ],
      }
    }
    onLayout={[Function]}
    onRequestChangeTab={[Function]}
    render={[Function]}
    style={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;
